bigbone/src/main/kotlin/social/bigbone/api/exception/BigBoneClientInstantiationException.kt

```diff
@@ -0,0 +1,47 @@
+package social.bigbone.api.exception
+
+import okhttp3.Response
+import social.bigbone.MastodonClient
+import social.bigbone.nodeinfo.entity.NodeInfo
+
+/**
+ * Exception thrown if we could not instantiate a [MastodonClient]. Mostly used to wrap other more specific exceptions.
+ */
+open class BigBoneClientInstantiationException : Exception {
```

Comment:
I might lack some Kotlin knowledge here, but couldn't we inherit from `RuntimeException` instead so that we do not need to explicitly handle or declare exceptions of type `BigBoneClientInstantiationException`? Looking at the modified samples in this PR, I see lots of checked exceptions wrapped in `RuntimeException`.

Reply:
> Looking at the modified samples in this PR, I see lots of checked exceptions wrapped in `RuntimeException`.

I did that less because I wanted to explicitly wrap it that way. I just wanted to do _something_ that would crash the sample in case we run into that error as I didn’t know which other path we should go with. I didn’t want to wrap the whole sample code (in same cases) in the `try` block.

That I rethrow a `RuntimeException` has no specific reason. 😊 

Extending `RuntimeException` or `IllegalStateException` (which itself extends `RuntimeException`) would still be an option though.

Reply:
I’ve replaced the try/catch blocks with an extended `throws` clause in the method header like we already did for `BigBoneRequestException` @andregasser

Reply:
Hello @PattaFeuFeu Thanks a lot for the changes. Let's assume we let `BigBoneClientInstantiationException` inherit from `RuntimeException` instead of `Exception`. Wouldn't this simplify library user's life as they a) do not need to declare the exception and b) also do not need to catch it if they do not want to?

Reply:
Kotlin does not have checked exceptions, so wouldn't this would only make a difference in Java? If we want that exception to be unchecked in Java as well, I think we could just remove the `@Throws` annotation - but that seems to be a design decision that we would want to discuss in general, outside of this PR.

For what it's worth, if we discuss the design of our exceptions, we could also talk about naming (which one's get prefixed with "BigBone" and which one's don't?), or whether it makes sense to let all of them inherit from some `BigBoneException` just to group all of them for easier catching.

Reply:
Hello @bocops Thanks for the feedback. As you mention, Kotlin does not care about the exception type. But I think it matters for Java devs. I think we could do Java devs a favor if we do not force them to declare/handle exceptions.

Let's say we do this:
- Base all the exceptions we throw on `RuntimeException`
- Remove all the `@Throws` annotations from our methods
- Wrap potential checked exceptions (IOException and such) in RuntimeException before we throw it

Doing so, Java devs should be happy, as they do not explicitly need to handle/declare our exceptions.

Please correct me if I am wrong.

Reply:
I believe we don't need to change our Exceptions to inherit from RuntimeException explicitly, or wrap other Exceptions, as long as we remove the `@Throws` annotations we currently have in place. That part would definitely be worth discussing, though. :)

Reply:
Hello @bocops, Thanks for the example code on Slack, which I quickly repost here for context:

Kotlin code in BigBone:
```kotlin
package social.bigbone.sample

import social.bigbone.api.exception.BigBoneRequestException
import java.io.IOException

class TestThrower {

    @Throws(BigBoneRequestException::class)
    fun throwDeclaredBigBoneException() {
        throw BigBoneRequestException("foo")
    }

    fun throwUndeclaredBigBoneException() {
        throw BigBoneRequestException("foo")
    }

    @Throws(IOException::class)
    fun throwDeclaredJavaException() {
        throw IOException("foo")
    }

    fun throwUndeclaredJavaException() {
        throw IOException("foo")
    }
}
```

Java call site:
![image](https://github.com/andregasser/bigbone/assets/535130/c342490c-6996-45d7-8aab-014aacd3a010)



Based on your results, I suggest, we remove the `@Throws` annotations on all our methods and continue to base our exceptions on simply `Exception` as it clearly works.

Would that be ok for all of you?

Reply:
> Would that be ok for all of you?

I don’t have a specific opinion on that. But in any case, I would not want to do all of that in this PR. At least not for the other Exceptions that already existed. I could rewrite the new Exceptions added here to not have the `@Throws` declaration (I would still add it to the kDoc though?).

Reply:
@PattaFeuFeu Fully agree with you. Thanks for the reply. 👍

Reply:
I have removed the `@Throws` annotations for _public_ methods but kept them for internal/private ones so that we know what to expect.

I have kept the kDoc `@throws` notes even for the public ones.